File: testdata/precision-suite-clojure/tp_sqlinjection.clj
(ns app.sqlinjection
  "SQL injection (CWE-89): a request value is interpolated into a SQL string via
   `str` concatenation and handed to clojure.java.jdbc/query. A correct scanner
   fires TAINT-001. (The parameterized [\"... ?\" v] vector form is the safe
   counterpart — see clean_safe_db.clj.)"
  (:require [clojure.java.jdbc :as jdbc]))

(def db {:dbtype "postgresql" :dbname "app"})

;; String-concatenated query: the tainted id is interpolated into the SQL text.
(defn find-user [req]
  (let [id (:params req)]
    (jdbc/query db (str "select * from users where id = " id)))) ; nox-expect: TAINT-001

;; execute! with a concatenated statement is equally injectable.
(defn delete-user [req]
  (let [uid (:query-string req)]
    (jdbc/execute! db (str "delete from users where id = " uid)))) ; nox-expect: TAINT-001

File: testdata/precision-suite-lua/tp_sqlinjection.lua
-- SQL injection (CWE-89): an untrusted value is concatenated into a SQL string
-- passed to a luasql connection:execute (no bind parameters) or an lsqlite3
-- db:exec. A correct scanner fires TAINT-001.

-- luasql: request argument interpolated into the SQL string.
local function lookup(conn)
  local id = ngx.req.get_uri_args().id
  return conn:execute("SELECT * FROM users WHERE id = " .. id) -- nox-expect: TAINT-001
end

-- lsqlite3: a POST argument interpolated into the query.
local function search(db)
  local q = ngx.req.get_post_args().q
  db:exec("SELECT * FROM items WHERE name LIKE '%" .. q .. "%'") -- nox-expect: TAINT-001
end
